url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/GantryAgent.java

```diff
@@ -0,0 +1,252 @@
+	}
+
+	@Override
+	public void msgRemoveBin(Bin bin) {
+		print("Received msgRemoveBin");
+		// synchronized(feeders){
+		for (MyFeeder currentFeeder : feeders) {
+			if (currentFeeder.state == FeederStatus.FULL && currentFeeder.requestedType.equals(bin.part.type)) {
+				currentFeeder.state = FeederStatus.PURGING;
+				break;
+			}
+		}
+		// }
+		bin.binState = BinStatus.DISCARDING;
+		stateChanged();
+	}
+
+	@Override
+	public void msgReceiveBinDone(Bin bin) {
+		print("Received msgReceiveBinDone from graphics");
+		bin.binState = BinStatus.OVER_FEEDER;
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgDropBinDone(Bin bin) {
+		print("Received msgdropBinDone from graphics");
+		bin.binState = BinStatus.EMPTY;
+		animation.release();
+		// waitForDrop = false;
+		stateChanged();
+	}
+
+	@Override
+	public void msgRemoveBinDone(Bin bin) {
+		print("Received msgremoveBinDone from graphics");
+		// binList.remove(bin);
+		bin.binState = BinStatus.FULL;
+		animation.release();
+		stateChanged();
+	}
+
+	// SCHEDULER
+	@Override
+	public boolean pickAndExecuteAnAction() {
+		synchronized(binList){
+			for (Bin bin : binList) {
+				if (bin.binState == BinStatus.PENDING) {
+					addBinToGraphics(bin);
+					return true;
+				}
+			}
+		}
+		for (MyFeeder currentFeeder : feeders) {
+			if (currentFeeder.state == FeederStatus.REQUESTED_PARTS) {
+				for (Bin bin : binList) {
+					if (bin.part.type.equals(currentFeeder.getRequestedType()) && bin.binState == BinStatus.FULL) {
+						moveToFeeder(bin, currentFeeder);
+						return true;
+					}
+				}
+			}
+		}
+
+		for (Bin bin : binList) {
+			if (bin.binState == BinStatus.DISCARDING) {
+				discardBin(bin);
+				return true;
+			}
+		}
+		return false;
+	}
+
+	// ACTIONS
+	public void moveToFeeder(Bin bin, MyFeeder feeder) {
+		print("Moving bin to over feeder " + feeder.feeder.toString());
+		bin.binState = BinStatus.MOVING;
+		feeder.state = FeederStatus.BEING_MOVED_TO;
+
+		gantryGraphics.receiveBin(bin, feeder.getFeeder());
+		// waitForDrop = true;
+		try {
+			animation.acquire();
+		} catch (InterruptedException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		fillFeeder(bin, feeder);
+
+		stateChanged();
+	}
+
+	public void fillFeeder(Bin bin, MyFeeder feeder) {
+		print("Placing bin in feeder and filling feeder " + feeder.feeder.toString());
+		bin.binState = BinStatus.FILLING_FEEDER;
+		feeder.state = FeederStatus.FULL;
+		// waitForDrop = false;
+
+		gantryGraphics.dropBin(bin, feeder.getFeeder());
+
+		try {
+			animation.acquire();
+		} catch (InterruptedException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		feeder.getFeeder().msgHereAreParts(bin.part.type, bin);
+
+		stateChanged();
+	}
+
+	public void discardBin(Bin bin) {
+		print("Discarding bin");
+		bin.binState = BinStatus.FULL;
+
+		gantryGraphics.removeBin(bin);
+
+		try {
+			animation.acquire();
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)